### ISSUE TITLE
Using TF hub encoder after fine-tuning

### ISSUE THREAD

**Issue by PrestonMWF** [tensorflow/hub#641]
I'm doing a downstream NLP prediction task and I've been able to fine-tune a pre-trained model (universal sentence encoder, specifically) that delivers a good result. However, I now want to use the fine-tuned model to encode new text with the updated weights so unseen text coming in can be better represented-- is this possible? If so, how does it work?

**Comment by akhorlin** [comment]
Hi Mark,

Is your concern that the model you trained will be used on a out of distribution text for inference are you looking to fine-tune USE in unsupervised or semi-supervised fashion on you own data? For the later, you can borrow the ideas for pretraining either from the original USE paper or from more recent approaches used for transformers like BERT and ALBERT. For example the TF Model Garden has canonical implementations of the recent transformer-based models: https://github.com/tensorflow/models/tree/master/official/nlp.

**Comment by PrestonMWF** [comment]
Hi Andrey,

The idea is USE creates reasonable sentence embeddings out of the box but, I want to update the model pre-train weights using a supervised task with my domain specific text. I've done this using TF and the hub.KerasLayer(..., trainable = True) approach with a final softmax layer for classification, which provides good prediction results. Following this the weights have been updated on a downstream task and now, I want to use them to encode other domain specific text, which obviously can't be done with the new TF model as there's no method for it. But, hub.load(...) models can do encoding and I'm wondering if I can somehow pass on weights from the fine-tuned model for new text encoding somehow. This way, new USE embeddings will be updated and more domain specific. Thanks for the quick response on the previous post as well!

**Comment by akhorlin** [comment]
You can try to export a piece of the model (before softmax layer) from the first stage of your pipeline using tf.keras.models.save_model() ([example](https://github.com/tensorflow/hub/blob/f8fac5d86dfb6b06ffa05bc237c736586fdafd4b/docs/tf2_saved_model.md#saving-from-keras) that shows how to save a piece of a model). The resulting SavedModel should have the same API as the original USE (text -> feature vector). Then, you have use resulting SavedModel with hub.load() api to generate feature vectors for the text in question, similar as your would the original USE.

**Comment by PrestonMWF** [comment]
Amazing, this works perfectly-- the example link you provided had exactly what I needed. I was able to transform new text after training with a softmax layer and my task is much better.  Thank you very much for the prompt response and insightful help!
